src/Suave/Logging/LogLine.fs

```diff
@@ -13,23 +15,45 @@ type LogLine =
     level         : LogLevel
     /// the source of the log line, e.g. 'ModuleName.FunctionName'
     path          : string
-    /// the message that the application wants to log
-    message       : string
+    /// the template message that the application wants to log
+    template      : string
+    /// a function which can format the log template message. When the
+    /// message is pre-formatted, this is None. When the log message 
+    /// contains structured data, this function can be used to render
+    /// the message template with the data.
+    formatter     : LogTemplateFormatter option
+    /// any key-value based data to log
+    data          : Map<string, obj>
     /// an optional exception
     ``exception`` : exn option
     /// timestamp when this log line was created
     tsUTCTicks    : int64 }
-    
+
+    /// Formats the log line as a string
+    member x.formatMessage () =
+        match x.formatter with
+        | None -> x.template
+        | Some fmter -> fmter x.template x.GetDataByName
+
+    /// Gets a data value
+    member x.GetDataByName name = Option.toObj (x.data.TryFind name)
+
 [<CompilationRepresentation(CompilationRepresentationFlags.ModuleSuffix)>]
 module LogLine =
   let mk path level trace ex message =
-    { message       = message
+    { template      = message
+      formatter     = None
+      data          = Map.empty
       level         = level
       path          = path
       ``exception`` = ex
       trace         = trace
       tsUTCTicks  = Globals.utcNow().Ticks }
 
+  /// assists with type inference, avoiding ugly 'box' calls for each
+  /// value item when constructing the log data map 
+  let mapLogData (data: (string*obj) list) = Map.ofList data
```

Comment:
OK most of this makes sense, at first glance.

Is there any reason why the property names ("instant" and "time" is your example) need to be specified twice? They are specified first in the template message, and again as a tuple via the `<+>` (Message.fieldWithUnit). Are you thinking that positional matching (like String.Format or Printf) is more brittle?